app/azure_logging.py

```diff
@@ -0,0 +1,61 @@
+import logging
+from typing import Optional
+
+from opencensus.ext.azure.log_exporter import AzureLogHandler
+from opencensus.trace import config_integration
+from opencensus.trace.samplers import AlwaysOnSampler
+from opencensus.trace.tracer import Tracer
+
+def telemetry_processor_callback_function(envelope):
+    envelope.tags['ai.cloud.role'] = 'model-api'
+    
+
+class ExceptionTracebackFilter(logging.Filter):
+    """
+    If a record contains 'exc_info', it will only show in the 'exceptions' section of Application Insights without showing
+    in the 'traces' section. In order to show it also in the 'traces' section, we need another log that does not contain 'exc_info'.
+    """
+    def filter(self, record):
+        if record.exc_info:
+            logger = logging.getLogger(record.name)
+            _, exception_value, _ = record.exc_info
+            message = f"{record.getMessage()}\nException message: '{exception_value}'"
+            logger.log(record.levelno, message)
+
+        return True
+
+
+def initialize_logging(logging_level: int, correlation_id: Optional[str] = None) -> logging.LoggerAdapter:
+    """
+    Adds the Application Insights handler for the root logger and sets the given logging level.
+    Creates and returns a logger adapter that integrates the correlation ID, if given, to the log messages.
+
+    :param logging_level: The logging level to set e.g., logging.WARNING.
+    :param correlation_id: Optional. The correlation ID that is passed on to the operation_Id in App Insights.
+    :returns: A newly created logger adapter.
+    """
+    logger = logging.getLogger()
+
+    try:
+        # picks up APPLICATIONINSIGHTS_CONNECTION_STRING automatically
+        azurelog_handler = AzureLogHandler()
+        azurelog_handler.add_telemetry_processor(telemetry_processor_callback_function)
+        azurelog_handler.addFilter(ExceptionTracebackFilter())
+        logger.addHandler(azurelog_handler)
+    except ValueError as e:
+        logger.error(f"Failed to set Application Insights logger handler: {e}")
```

Comment:
What situations is it okay for the logger handler to fail? Not sure I would catch this exception

Reply:
I'm not convinced I should get an error when nothing is wrong i.e. locally when no app insights is set
This will also silently fail if `APPLICATIONINSIGHTS_CONNECTION_STRING` isn't set in a deployment environment. Think if I couldn't see the logs I'd want that to crash the app(?)

Reply:
yeah...both approaches are fine, but happy to switch it to fail, but that'll also require us to special-case local dev

if it reasonable to say, if the env var exists, then `raise` 
otherwise, carry on

i.e. take the existence of `APPLICATIONINSIGHTS_CONNECTION_STRING` as a strong signal that App Insights logging should work, and that failure is not an option. BUT, that if it's *not* set at all, we're OK to carry on.

This keeps the logic of "do I raise an error or not" in here, rather than having to pass stuff into this function to say "it's OK to fail when running local"

Reply:
(OR, is there already something that signals "you're running locally" that we can inspect?)

Reply:
agree. did [suggest](https://github.com/UCLH-Foundry/Dash-Seedling/pull/10#pullrequestreview-1393931148) using the set `ENVIRONMENT` variable for local and returning early. thoughts?

Reply:
Thing is, if you're testing or developing the logging, you won't want it to return early. You'll want it to log to App Insights.

Suggest...

- attempt to wire up App Insights (as this is the default, and most likely/important path)
- in the `except` block `raise` (i.e. re-throw) the exception in any situation *other than* when `ENVIRONMENT` is local
  - if it fails in the local env (which it will most of the time, because the env var won't be set) that's fine

Reply:
My assumption was `ENVIRONMENT=local` means no cloud deployment. Guess that's not clear 😞

Reply:
that's clear to me
not sure what you mean

Reply:
My suggestion was that we should continue to wire up app insights logging in **all** situations

The only change is then how we deal with failure to do that wire up:
- we'd re-raise the exception on failure to wire up app insights in all situations **unless** we are local
- when local, we would not see the failure as terminal

I think you're suggesting "don't even attempt to wire it up when local". 

If that's the case, I'm not keen on that approach because there are situations where you want it wired up locally (any bugs or enhancements to the logging itself). Plus, you end up with bigger code branches of "stuff that only runs locally" and "stuff that runs in the cloud".

Reply:
I agree with Martin.
In his suggested approach, most of the logging related code for local and other environment will be same. 
And code to treat the raised exception depends on the environment 
instead of code to raise the exception itself.
On local, to me conceptually, it is like a warning; the developer can decide what to do with it.

Reply:
this is what I'd suggest: https://github.com/UCLH-Foundry/Dash-Seedling/blob/6c17acd62b1ff9edabb2232b3baaf4b7769a122d/app/log.py (also fixed the line lengths)

exceptions should not be expected, in my opinion at least. Saying that, I'm happy to agree to disagree and to reraise only in any non-local environments

Reply:
> On local, to me conceptually, it is like a warning; the developer can decide what to do with it.

I don't think there should be a warning if the behaviour is entirely expected

Reply:
I think there can be warning raised if we know the behaviour 
but not the intentions of the user (developer in this case).

But with this much discussion going on here, happy either way :)

Reply:
Did it this way 264ae7850adba0cdd2f517af878896f6daedf171